react-app/src/components/ValidatorDetailScreen/ValidatorDetailScreenAPI.tsx

```diff
@@ -0,0 +1,103 @@
+  RequestStateLoaded,
+  RequestStateLoading,
+} from "../../models/RequestState";
+import { useStakingAPI } from "../../api/stakingAPI";
+import { ConnectionStatus, useWallet } from "../../providers/WalletProvider";
+import { useDistributionAPI } from "../../api/distributionAPI";
+import { BigNumberCoin } from "../../models/coin";
+import { useSlashingAPI } from "../../api/slashingAPI";
+import { Validator as ValidatorCosmos } from "../../models/cosmos/staking";
+import { Validator, YourStake } from "./ValidatorDetailScreenModel";
+
+interface ValidatorQuery {
+  (): {
+    fetch: (validatorAddress: string) => Promise<void>;
+    requestState: RequestState<Validator>;
+  };
+}
+
+export const useValidatorQuery: ValidatorQuery = () => {
+  const [requestState, setRequestState] =
+    useState<RequestState<Validator>>(RequestStateInitial);
+
+  const wallet = useWallet();
+  const stakingAPI = useStakingAPI();
+  const slashingAPI = useSlashingAPI();
+  const distributionAPI = useDistributionAPI();
+
+  const fetch = useCallback(
+    // eslint-disable-next-line complexity
+    async (validatorAddress: string) => {
+      setRequestState(RequestStateLoading);
+      try {
+        const promises: Promise<
+          // eslint-disable-next-line @typescript-eslint/no-redundant-type-constituents
+          | ValidatorCosmos
+          | number
+          | Omit<YourStake, "reward">
+          | BigNumberCoin
+          | null
+        >[] = [
+          stakingAPI.getValidator(validatorAddress),
+          distributionAPI.getValidatorExpectedReturn(validatorAddress),
+        ];
+
+        if (wallet.status === ConnectionStatus.Connected) {
+          promises.push(
+            stakingAPI.getDelegation(wallet.account.address, validatorAddress)
+          );
+          promises.push(
+            distributionAPI.getDelegationRewardsByValidator(
+              wallet.account.address,
+              validatorAddress
+            )
+          );
+        }
+
+        const [validatorUntyped, expectedReturn, ...restData] =
+          await Promise.all(promises);
+        const validator = validatorUntyped as ValidatorCosmos;
+
+        const [signingInfo, uptime] = await Promise.all([
+          slashingAPI.getSigningInfo(validator.consensusPubAddr),
+          slashingAPI.getValidatorUptime(validator.consensusPubAddr),
+        ]);
+
+        // stake will be nullable when
+        const stake =
+          // sanity check
+          restData.length > 0 &&
+          // 1. no stake for this validator
+          restData[0] != null &&
+          // 2. wallet not connected
+          wallet.status === ConnectionStatus.Connected
+            ? {
+                ...(restData[0] as YourStake),
+                reward: restData[1] as BigNumberCoin,
+              }
+            : null;
+
+        const res = {
+          ...validator,
+          uptime,
+          stake,
+          signingInfo,
+          expectedReturn: expectedReturn as number,
+        };
```

Comment:
kinda messy to merge everything together, imo returning stake and validator separately would be clearer